### PR TITLE
UI-18914: Adds ssh keys data source, volumes data source

### PR DESCRIPTION
**Pull request by BShrivastav**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000
https://jiracloud.swg.usma.ibm.com:8443/browse/UI-18914

Output from acceptance testing:

```

$ make testacc TEST=./ibm/service/vpc TESTARGS='-run=TestAccIBMIsSshKeysDataSourceBasic'
=== RUN   TestAccIBMIsSshKeysDataSourceBasic
--- PASS: TestAccIBMIsSshKeysDataSourceBasic (36.26s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/vpc	37.838s
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm	69.128s

$ make testacc TEST=./ibm/service/vpc TESTARGS='-run=TestAccIBMIsVolumesDataSourceBasic'
=== RUN   TestAccIBMIsVolumesDataSourceBasic
--- PASS: TestAccIBMIsVolumesDataSourceBasic (86.67s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/vpc	87.632s

...
```
[Responses.txt](https://github.com/ibm-vpc/terraform-provider-ibm/files/7512682/Responses.txt)

### CODE REVIEW

File: examples/ibm-is-ng/main.tf [hunk 1]
```diff
@@ -675,3 +675,11 @@ data "ibm_is_placement_groups" "is_placement_groups" {
 ## List regions 
 data "ibm_is_regions" "regions" {
 }
+
+# List ssh keys 
+data "ibm_is_ssh_keys" "example-keys" {
```
Comment (ujjwal-ibm) [inline examples/ibm-is-ng/main.tf:680]: just example is fine

File: examples/ibm-is-ng/main.tf [hunk 2]
```diff
@@ -675,3 +675,11 @@ data "ibm_is_placement_groups" "is_placement_groups" {
...
+}
+
+# List volumes
+data "ibm_is_volumes" "example-volumes" {
```
Comment (ujjwal-ibm) [inline examples/ibm-is-ng/main.tf:684]: same as above

File: ibm/data_source_ibm_is_ssh_keys.go [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	// "github.ibm.com/ibmcloud/vpc-go-sdk/vpcv1"
```
Comment (ujjwal-ibm) [inline ibm/data_source_ibm_is_ssh_keys.go:15]: remove commented code

File: ibm/data_source_ibm_is_ssh_keys.go [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+	isKeysLimit            = "limit"
+	isKeysTotalCount       = "total_count"
+)
+
+func dataSourceIBMIsSshKeys() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: dataSourceIBMIsSshKeysRead,
+
+		Schema: map[string]*schema.Schema{
+			"first": &schema.Schema{
```
Comment (ujjwal-ibm) [inline ibm/data_source_ibm_is_ssh_keys.go:37]: is this needed ?

File: ibm/data_source_ibm_is_ssh_keys.go [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+						},
+						isKeyType: &schema.Schema{
+							Type:        schema.TypeString,
+							Computed:    true,
+							Description: "The crypto-system used by this key.",
+						},
+					},
+				},
+			},
+			isKeysLimit: &schema.Schema{
```
Comment (ujjwal-ibm) [inline ibm/data_source_ibm_is_ssh_keys.go:129]: this also isnt needed. Check once with other resources

File: ibm/data_source_ibm_is_ssh_keys.go [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+				Type:        schema.TypeInt,
+				Computed:    true,
+				Description: "The maximum number of resources that can be returned by the request.",
+			},
+			"next": &schema.Schema{
```
Comment (ujjwal-ibm) [inline ibm/data_source_ibm_is_ssh_keys.go:134]: same, not needed

File: ibm/data_source_ibm_is_ssh_keys.go [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+func dataSourceIBMIsSshKeysID(d *schema.ResourceData) string {
+	return time.Now().UTC().String()
+}
+
+// func dataSourceKeyCollectionFlattenFirst(result vpcv1.KeyCollectionFirst) (finalList []map[string]interface{}) {
+// 	finalList = []map[string]interface{}{}
+// 	finalMap := dataSourceKeyCollectionFirstToMap(result)
+// 	finalList = append(finalList, finalMap)
+
+// 	return finalList
```
Comment (ujjwal-ibm) [inline ibm/data_source_ibm_is_ssh_keys.go:215]: remove commented code

File: ibm/data_source_ibm_is_ssh_keys.go [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+}
+
+// func dataSourceKeyCollectionFlattenNext(result vpcv1.KeyCollectionNext) (finalList []map[string]interface{}) {
+// 	finalList = []map[string]interface{}{}
+// 	finalMap := dataSourceKeyCollectionNextToMap(result)
+// 	finalList = append(finalList, finalMap)
+
+// 	return finalList
+// }
+
```
Comment (ujjwal-ibm) [inline ibm/data_source_ibm_is_ssh_keys.go:299]: remove commented code

File: ibm/data_source_ibm_is_ssh_keys_test.go [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+func TestAccIBMIsSshKeysDataSourceBasic(t *testing.T) {
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: testAccCheckIBMIsSshKeysDataSourceConfigBasic(),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttrSet("data.ibm_is_ssh_keys.is_ssh_keys", "id"),
+					// resource.TestCheckResourceAttrSet("data.ibm_is_ssh_keys.is_ssh_keys", "first.#"),
```
Comment (ujjwal-ibm) [inline ibm/data_source_ibm_is_ssh_keys_test.go:22]: remove commented code

File: ibm/data_source_ibm_is_volumes.go [hunk 10]
```diff
@@ -0,0 +1,968 @@
...
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	// "github.ibm.com/ibmcloud/vpc-go-sdk/vpcv1"
```
Comment (ujjwal-ibm) [inline ibm/data_source_ibm_is_volumes.go:15]: remove

File: ibm/data_source_ibm_is_volumes.go [hunk 11]
```diff
@@ -0,0 +1,968 @@
...
+	isVolumesZoneHref                                      = "href"
+	isVolumesZoneName                                      = "name"
+)
+
+func dataSourceIBMIsVolumes() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: dataSourceIBMIsVolumesRead,
+
+		Schema: map[string]*schema.Schema{
+			"first": &schema.Schema{
```
Comment (ujjwal-ibm) [inline ibm/data_source_ibm_is_volumes.go:91]: same as above

File: ibm/data_source_ibm_is_volumes_test.go [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+func TestAccIBMIsVolumesDataSourceBasic(t *testing.T) {
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: testAccCheckIBMIsVolumesDataSourceConfigBasic(),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttrSet("data.ibm_is_volumes.is_volumes", "id"),
+					// resource.TestCheckResourceAttrSet("data.ibm_is_volumes.is_volumes", "first.#"),
```
Comment (ujjwal-ibm) [inline ibm/data_source_ibm_is_volumes_test.go:22]: same as above

File: website/docs/d/is_ssh_keys.html.markdown [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+---
+
+# ibm_is_ssh_keys
+
+Provides a read-only data source for KeyCollection. You can then reference the fields of the data source in other resources within the same configuration using interpolation syntax.
+
+## Example Usage
+
+```hcl
+data "ibm_is_ssh_keys" "example" {
```
Comment (ujjwal-ibm) [inline website/docs/d/is_ssh_keys.html.markdown:16]: add example for filters as well

File: website/docs/d/is_ssh_keys.html.markdown [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+}
+```
+
+
+## Attribute Reference
+
+In addition to all argument references listed, you can access the following attribute references after your data source is created.
+
+* `id` - The unique identifier of the KeyCollection.
```
Comment (ujjwal-ibm) [inline website/docs/d/is_ssh_keys.html.markdown:25]: change * to -

File: website/docs/d/is_ssh_keys.html.markdown [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+	* `fingerprint` - (Required, String) The fingerprint for this key.  The value is returned base64-encoded and prefixed with the hash algorithm (always `SHA256`).
+	* `href` - (Required, String) The URL for this key.
+	  * Constraints: The value must match regular expression `/^http(s)?:\/\/([^\/?#]*)([^?#]*)(\\?([^#]*))?(#(.*))?$/`.
+	* `id` - (Required, String) The unique identifier for this key.
+	  * Constraints: The maximum length is `64` characters. The minimum length is `1` character. The value must match regular expression `/^[-0-9a-z_]+$/`.
+	* `length` - (Required, Integer) The length of this key (in bits).
+	  * Constraints: Allowable values are: `2048`, `4096`.
+	* `name` - (Required, String) The unique user-defined name for this key. If unspecified, the name will be a hyphenated list of randomly-selected words.
+	  * Constraints: The maximum length is `63` characters. The minimum length is `1` character. The value must match regular expression `/^([a-z]|[a-z][-a-z0-9]*[a-z0-9])$/`.
+	* `public_key` - (Required, String) The public SSH key, consisting of two space-separated fields: the algorithm name, and the base64-encoded key.
```
Comment (ujjwal-ibm) [inline website/docs/d/is_ssh_keys.html.markdown:44]: remove required/optional from exported attributes

File: website/docs/d/is_volumes.html.markdown [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+data "ibm_is_volumes" "example" {
+}
+```
+
+
+## Attribute Reference
+
+In addition to all argument references listed, you can access the following attribute references after your data source is created.
+
+* `id` - The unique identifier of the VolumeCollection.
```
Comment (ujjwal-ibm) [inline website/docs/d/is_volumes.html.markdown:25]: same as other doc, replace * with - and remove required/optional from exported attributes

**Review by ujjwal-ibm — CHANGES_REQUESTED** [review]

File: ibm/data_source_ibm_is_ssh_keys.go [hunk 17]
```diff
@@ -0,0 +1,319 @@
...
+	// filter for resource group Id
+	resourceGroupId := d.Get("resource_group").(string)
+
+	for {
+		listKeysOptions := &vpcv1.ListKeysOptions{}
+		if start != "" {
+			listKeysOptions.Start = &start
+		}
+		//filter
+		if listKeysOptions.ResourceGroupID != nil {
```
Comment (deepaksibm) [inline ibm/data_source_ibm_is_ssh_keys.go:179]: if resourceGroupId != "" {

File: website/docs/d/is_ssh_keys.html.markdown (hunk 14 above)
Comment (BShrivastav) [reply website/docs/d/is_ssh_keys.html.markdown:25]: all the places, or just the bullets ?

File: ibm/data_source_ibm_is_volumes.go [hunk 18]
```diff
@@ -0,0 +1,897 @@
...
+	return &schema.Resource{
+		ReadContext: dataSourceIBMIsVolumesRead,
+
+		Schema: map[string]*schema.Schema{
+			"volume_name": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Volume name identifier.",
+			},
+			"zone_name": &schema.Schema{
```
Comment (ujjwal-ibm) [inline ibm/data_source_ibm_is_volumes.go:95]: make it zone

File: ibm/data_source_ibm_is_volumes_test.go [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+					resource.TestCheckResourceAttrSet("data.ibm_is_volumes.is_volumes", "volumes.#"),
+				),
+			},
+		},
+	})
+}
+
+func testAccCheckIBMIsVolumesDataSourceConfigBasic() string {
+	return fmt.Sprintf(`
+		data "ibm_is_volumes" "is_volumes" {
```
Comment (ujjwal-ibm) [inline ibm/data_source_ibm_is_volumes_test.go:31]: add a scenario for filters

File: website/docs/d/is_volumes.html.markdown [hunk 20]
```diff
@@ -0,0 +1,154 @@
...
+			- `id` - (String) The unique identifier for this virtual server instance.
+			  - Constraints: The maximum length is `64` characters. The minimum length is `1` character. The value must match regular expression `/^[-0-9a-z_]+$/`.
+			- `name` - (String) The user-defined name for this virtual server instance (and default system hostname).
+			  - Constraints: The maximum length is `63` characters. The minimum length is `1` character. The value must match regular expression `/^-?([a-z]|[a-z][-a-z0-9]*[a-z0-9]|[0-9][-a-z0-9]*([a-z]|[-a-z][-a-z0-9]*[a-z0-9]))$/`.
+		- `name` - (String) The user-defined name for this volume attachment.
+		  - Constraints: The maximum length is `63` characters. The minimum length is `1` character. The value must match regular expression `/^-?([a-z]|[a-z][-a-z0-9]*[a-z0-9]|[0-9][-a-z0-9]*([a-z]|[-a-z][-a-z0-9]*[a-z0-9]))$/`.
+		- `type` - (String) The type of volume attachment.
+		  - Constraints: Allowable values are: `boot`, `data`.
+	- `zone` - (List) The zone this volume resides in.
+	Nested scheme for **zone**:
```
Comment (ujjwal-ibm) [inline website/docs/d/is_volumes.html.markdown:149]: zone means zone name. No need to create a list out of it

**Review by ujjwal-ibm — CHANGES_REQUESTED** [review]

File: website/docs/d/is_volumes.html.markdown (hunk 20 above)
Comment (BShrivastav) [reply website/docs/d/is_volumes.html.markdown:136]: done.

File: ibm/data_source_ibm_is_volumes_test.go (hunk 19 above)
Comment (BShrivastav) [reply ibm/data_source_ibm_is_volumes_test.go:45]: done.

File: ibm/data_source_ibm_is_volumes.go (hunk 18 above)
Comment (BShrivastav) [reply ibm/data_source_ibm_is_volumes.go:95]: keeping zone_name to avoid redundancy.

**Comment by astha-jain** [comment]
is this merged?

**Comment by astha-jain** [comment]
@BShrivastav What is the status of this task?

**Comment by BShrivastav** [comment]
https://github.com/IBM-Cloud/terraform-provider-ibm/pull/3694 raised Upstream, closing internal.

**Pull request closed without merging**